### PR TITLE
GET RPC requests

### PR DESCRIPTION
**Pull request by stroop23**
These changes make it possible to do get requests without any nginx+lua rewriting.
If the "body" parameter is passed, it will be treated as containing url encoded valid JSON.
If no body parameter is passed, a request for the definition of the particular method as assumed when enabled through setting the piqi_rpc/return_definition environment parameter to true.

### CODE REVIEW

File: src/piqi_rpc_webmachine_resource.erl [hunk 1]
```diff
@@ -140,12 +130,24 @@ known_content_type(ReqData, Context) ->
 
     % NOTE: Webmachine returns <<>> as a body even if "Content-Length" header is
     % not present.
-    Body = wrq:req_body(ReqData),
+    Body = case wrq:method(ReqData) of
+        'GET' -> 
+            Qs = wrq:req_qs(ReqData),
+            case proplists:get_value("body", Qs) of 
+                undefined -> <<>>;
+                BodyParm -> list_to_binary(BodyParm)
```
Comment (stroop23) [inline src/piqi_rpc_webmachine_resource.erl:138]: Is this safe? I don't know what webmachine/mochiweb do with regards to encodings.

Comment (yfyf) [inline src/piqi_rpc_webmachine_resource.erl:138]: No. Why are you doing this anyway?

Comment (stroop23) [inline src/piqi_rpc_webmachine_resource.erl:138]: piqi requires binary, webmachine delivers list and i do not like to use the elibs_types() stuff here.

Comment (yfyf) [inline src/piqi_rpc_webmachine_resource.erl:138]: Sorry, misunderstood you there, I thought you are converting a binary into a list. I that case, I think it's ok.

Comment (stroop23) [inline src/piqi_rpc_webmachine_resource.erl:138]: Google told me webmachine/mochiweb to list_to_binary/binary_to_list everywhere, and do not use unicode:\* so this is correct.

**Comment by yfyf** [comment]
why do we need this?

File: src/piqi_rpc_webmachine_resource.erl [hunk 2]
```diff
@@ -107,19 +107,9 @@ allowed_methods(ReqData, Context) ->
...
-        case wrq:method(ReqData) of
-            'POST' ->
-                FuncName = wrq:disp_path(ReqData),
-                lists:member($/, FuncName) andalso
-                    throw_malformed("invalid function name: " ++ FuncName);
-            _ -> ok
-        end,
+        FuncName = wrq:disp_path(ReqData),
+        lists:member($/, FuncName) andalso
+            throw_malformed("invalid function name: " ++ FuncName),
```
Comment (yfyf) [inline src/piqi_rpc_webmachine_resource.erl:112]: woah, what hacks, haha.

File: src/piqi_rpc_webmachine_resource.erl [hunk 3]
```diff
@@ -140,12 +130,24 @@ known_content_type(ReqData, Context) ->
...
     % not present.
-    Body = wrq:req_body(ReqData),
+    Body = case wrq:method(ReqData) of
+        'GET' -> 
+            Qs = wrq:req_qs(ReqData),
+            case proplists:get_value("body", Qs) of 
+                undefined -> <<>>;
+                BodyParm -> list_to_binary(BodyParm)
+            end;
+        'POST' ->
```
Comment (yfyf) [inline src/piqi_rpc_webmachine_resource.erl]: why not collapse these last two cases into `wrq:req_body(ReqData)` to preserve original semantics for all other requests?

Comment (yfyf) [inline src/piqi_rpc_webmachine_resource.erl]: e.g. in the future we might have `PUT` requests or whatever, which might have a body too.

File: src/piqi_rpc_webmachine_resource.erl [hunk 4]
```diff
@@ -178,18 +180,25 @@ known_content_type(ReqData, Context) ->
...
         case wrq:method(ReqData) of
-            'GET' -> % get Piqi: by default return it in Piq format
+            'GET' when Body =:= 'undefined' -> % get Piqi: by default return it in Piq format
                 [
                     {"text/plain", get_piqi_piq},
                     {"application/json", get_piqi_json},
                     {"application/xml", get_piqi_xml},
                     {"application/x-protobuf", get_piqi_pb}
                 ];
+            'GET' ->
```
Comment (yfyf) [inline src/piqi_rpc_webmachine_resource.erl:193]: GET requests should not have a Content-Type at all, because they don't have request entities.

Comment (yfyf) [inline src/piqi_rpc_webmachine_resource.erl:193]: I think you should instead pass a separate field which specifies the body type and then an Accept header to specify the return type expected.

**Comment by stroop23** [comment]
In response to most comments:
There are quite some inconsistencies/"bugs" in piqi-rpc itself. I'm not attempting to fix any of those. 
Least amount of change is the goal here. Any major bugfixing/refactoring must/will be done in the rewrite using cowboy.

File: src/piqi_rpc_webmachine_resource.erl (hunk 3 above)
Comment (stroop23) [inline src/piqi_rpc_webmachine_resource.erl]: I'll change the 'POST' into _, that will maintain old behaviour for anything besides GET.

**Comment by yfyf** [comment]
True, but before only `POST` requests were possible (for any real usage), `GET` requests were there only for "discoverability" reasons, now `GET` requests become a valid type of a request, so you might want to handle them cleanly.

But anyway, maybe you're right,  the `Content-Type` vs. `Accept` header issue and etc is probably best dealt with when webmachine is removed.

**Pull request merged**